flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/KNN.scala

```diff
@@ -0,0 +1,204 @@
+    */
+  def setBlocks(n: Int): KNN = {
+    require(n > 1, "Number of blocks must be positive.")
+    parameters.add(Blocks, n)
+    this
+  }
+}
+
+object KNN {
+
+  case object K extends Parameter[Int] {
+    val defaultValue: Option[Int] = None
+  }
+
+  case object DistanceMetric extends Parameter[DistanceMetric] {
+    val defaultValue: Option[DistanceMetric] = Some(EuclideanDistanceMetric())
+  }
+
+  case object Blocks extends Parameter[Int] {
+    val defaultValue: Option[Int] = None
+  }
+
+  def apply(): KNN = {
+    new KNN()
+  }
+
+  /** [[FitOperation]] which trains a KNN based on the given training data set.
+    * @tparam T Subtype of [[Vector]]
+    */
+  implicit def fitKNN[T <: Vector : TypeInformation] = new FitOperation[KNN, T] {
+    override def fit(
+        instance: KNN,
+        fitParameters: ParameterMap,
+        input: DataSet[T]): Unit = {
+      val resultParameters = instance.parameters ++ fitParameters
+
+      require(resultParameters.get(K).isDefined, "K is needed for calculation")
+
+      val blocks = resultParameters.get(Blocks).getOrElse(input.getParallelism)
+      val partitioner = FlinkMLTools.ModuloKeyPartitioner
+      val inputAsVector = input.asInstanceOf[DataSet[Vector]]
+
+      instance.trainingSet = Some(FlinkMLTools.block(inputAsVector, blocks, Some(partitioner)))
+    }
+  }
+
+  /** [[PredictDataSetOperation]] which calculates k-nearest neighbors of the given testing data
+    * set.
+    * @tparam T Subtype of [[Vector]]
+    * @return The given testing data set with k-nearest neighbors
+    */
+  implicit def predictValues[T <: Vector : ClassTag : TypeInformation] = {
+    new PredictDataSetOperation[KNN, T, (Vector, Array[Vector])] {
+      override def predictDataSet(
+          instance: KNN,
+          predictParameters: ParameterMap,
+          input: DataSet[T]): DataSet[(Vector, Array[Vector])] = {
+        val resultParameters = instance.parameters ++ predictParameters
+
+        instance.trainingSet match {
+          case Some(trainingSet) => {
+            val k = resultParameters.get(K).get
+            val blocks = resultParameters.get(Blocks).getOrElse(input.getParallelism)
+            val metric = resultParameters.get(DistanceMetric).get
+            val partitioner = FlinkMLTools.ModuloKeyPartitioner
+
+            // attach unique id for each data
+            val inputWithIndex: DataSet[(Long, T)] = input.zipWithIndex
+
+            // split data into multiple blocks
+            val inputSplit = FlinkMLTools.block(inputWithIndex, blocks, Some(partitioner))
+
+            // join input and training set
+            val crossed = trainingSet.cross(inputSplit).mapPartition {
+              (iter, out: Collector[(Vector, Vector, Long, Double)]) => {
+                for ((training, testing) <- iter) {
+                  for (a <- testing.values; b <- training.values) {
+                    // (training vector, input key, input vector, distance)
```

Comment:
Comment does not conform to code, should be: 

// (training vector, input vector, input key, distance)